browser/ui/views/toolbar/brave_toolbar_view.cc

```diff
@@ -0,0 +1,206 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/toolbar/brave_toolbar_view.h"
+
+#include "brave/browser/ui/views/toolbar/bookmark_button.h"
+#include "chrome/browser/ui/views/toolbar/toolbar_view.h"
+#include "chrome/browser/extensions/extension_util.h"
+#include "chrome/browser/ui/bookmarks/bookmark_bubble_sign_in_delegate.h"
+#include "chrome/browser/ui/layout_constants.h"
+#include "chrome/browser/ui/views/bookmarks/bookmark_bubble_view.h"
+#include "chrome/browser/ui/views/toolbar/browser_app_menu_button.h"
+#include "chrome/browser/ui/views/toolbar/home_button.h"
+#include "chrome/browser/ui/views/toolbar/reload_button.h"
+#include "chrome/browser/ui/views/toolbar/toolbar_button.h"
+#include "ui/base/material_design/material_design_controller.h"
+
+namespace {
+
+int GetToolbarHorizontalPadding() {
+  // In the touch-optimized UI, we don't use any horizontal paddings; the back
+  // button starts from the beginning of the view, and the app menu button ends
+  // at the end of the view.
+  return ui::MaterialDesignController::IsTouchOptimizedUiEnabled() ? 0 : 8;
+}
+
+}  // namespace
+
+void BraveToolbarView::Init() {
+  ToolbarView::Init();
+
+  // Only location bar in non-normal mode
+  if (!is_display_mode_normal()) {
+    return;
+  }
+
+  bookmark_ = new BookmarkButton(this);
+  bookmark_->set_triggerable_event_flags(ui::EF_LEFT_MOUSE_BUTTON | ui::EF_MIDDLE_MOUSE_BUTTON);
+  bookmark_->Init();
+  AddChildView(bookmark_);
+}
+
+void BraveToolbarView::ShowBookmarkBubble(
+    const GURL& url,
+    bool already_bookmarked,
+    bookmarks::BookmarkBubbleObserver* observer) {
+  // Show BookmarkBubble attached to Brave's bookmark button
+  // or the location bar if there is no bookmark button
+  // (i.e. in non-normal display mode).
+  views::View* anchor_view = location_bar();
+  BookmarkButton* const star_view = bookmark_button();
+  if (star_view && star_view->visible())
```

Comment:
nit: `star_view->visible()` can be removed because it is always visible.

Reply:
no longer true since `visible` is `false` when bookmarks are disabled.